File: src/solutions/java/MaximumProductSubarray.java
package solutions.java;

/*
 # Problem
 	- `Link`: https://leetcode.com/problems/maximum-product-subarray/
 # Solution
 	- `Author`: Kyungtaek Lim (Jonas)
 	- `Date`: Apr 8, 2025
 	- `Answer`: maxProduct
 */

public class MaximumProductSubarray {
	
	public static void main(String[] args) {
		
		System.out.println(maxProduct(new int[] {2,3,-2,4}));
		System.out.println(maxProduct(new int[] {-2,0,-1}));
		System.out.println(maxProduct(new int[] {3,-1,4}));
	}

	/*
	 # Option #1 
     - O(n)
	 */
    public static int maxProduct(int[] nums) {
    	
    	int maxVal = nums[0];
    	int minVal = maxVal;
    	int result = maxVal;
    	
    	for (int i = 1; i < nums.length; i++) {
    		int num = nums[i];
    		
    		if (num < 0) {
    			int temp = maxVal;
    			maxVal = minVal;
    			minVal = temp;
    		}
    		
    		maxVal = Math.max(num, maxVal * num);
    		minVal = Math.min(num, minVal * num);
    		
    		result = Math.max(result, maxVal);
    	}
    	
    	return result;
    }
}

File: src/solutions/typescript/MaximumProductSubarray.ts
/*
 # Problem
 	- `Link`: https://leetcode.com/problems/maximum-product-subarray/
 # Solution
 	- `Author`: Kyungtaek Lim (Jonas)
 	- `Date`: May 28, 2025
 	- `Answer`: maxProduct / maxProductAdvanced
*/


/*
    # Option #1
    - O(n)
*/
function maxProduct(nums: number[]): number {
    let min: number = nums[0], max: number = nums[0];
    let result: number = nums[0];
    for (const num of nums.slice(1)) {
        let temp: number = min;
        min = Math.min(min * num, num, max * num);
        max = Math.max(max * num, num, temp * num);
        result = Math.max(result, max);
    }
    return result;
};


/*
    # Option #2
    - O(n)
*/
function maxProductAdvanced(nums: number[]): number {
    let min: number = nums[0], max: number = nums[0];
    let result: number = nums[0];
    for (const num of nums.slice(1)) {
        let temp: number = min;
        if (num < 0) {
            min = Math.min(num, max * num);
            max = Math.max(num, temp * num);
        } else {
            min = Math.min(min * num, num);
            max = Math.max(max * num, num);
        }
        result = Math.max(result, max);
    }
    return result;
}